doc/md: sqlc guide (versioned)

crossworth opened the pull request.

rotemtam commented on doc/md/guides/frameworks/sqlc-versioned.md:
```diff
@@ -0,0 +1,193 @@
+
+This strategy brings the best of both worlds, we can have a single source of truth (our desired state) and have all the
+authored SQL files that will be used to reach the desired state.
+
```
```suggestion

To learn more about this approach and see it in action, watch this [YouTube video](https://www.youtube.com/watch?v=L-UlkXtp3OY)

```

rotemtam reviewed (commented):
good job @crossworth 
as with the last guide, lets incorporate a simple demonstration of how this interacts with the `sqlc` command. Show how you configure sqlc to use the migration dir as its source, and demo how you can update the schema and query, generate a migration and then use it.

crossworth commented:
Hi, I have added a few more "parts" to it, overall I re-structured to ensure that the concepts are presented first and focus that in this guide, while is called "Versioned migrations", we are using Versioned Migrations + Migration Authoring.

I added a little more of "code" to it as well, showing the generated code, sqlc commands and files, but I feel like the middle of the guide is almost `codeblocks` only, so feel free to propose changes to remove redundant parts or that makes no sense.

I tried to show more about sqlc, but it is really simple to use and even sqlc getting started is quite lean.

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
```diff
@@ -0,0 +1,285 @@
+slug: /guides/frameworks/sqlc-versioned
+---
+
+In our [previous sqlc guide](/guides/frameworks/sqlc-declarative) we saw how we can use Atlas to handle the schema
```
```suggestion
In our [previous sqlc guide](/guides/frameworks/sqlc-declarative), we saw how we can use Atlas to handle the schema
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
```diff
@@ -0,0 +1,285 @@
+and replayed on the database as needed. This ensures that the code to be executed on the database is known and approved
+beforehand and gives engineers ultimate control on exactly what code is going to be run in production.
+
+In this guide we will see how we can use sqlc with Atlas, using the versioned migration's workflow.
```
```suggestion
In this guide we will see how we can use sqlc with Atlas, using the versioned migrations workflow.
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
```diff
@@ -0,0 +1,285 @@
+In this guide we will see how we can use sqlc with Atlas, using the versioned migration's workflow.
+
+:::note
+If you haven't already done so, it's helpful if you have read the
```
```suggestion
If you haven't already done so, it will be helpful to have read the
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
```diff
@@ -0,0 +1,285 @@
+previous [sqlc guide](/guides/frameworks/sqlc-declarative) where we have explained some concepts that we will expand
+on in this guide. You should also check
+the [sqlc getting started with PostgreSQL](https://docs.sqlc.dev/en/latest/tutorials/getting-started-postgresql.html)
+guide since this guide continues where the sqlc guide left off.
```
```suggestion
guide since this guide continues where it left off.
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
```diff
@@ -0,0 +1,285 @@
+upgrade the schema from the current version to the next. Many schema management tools, such as Flyway and Liquibase,
+support this strategy and Atlas supports it as well.
+
+Atlas's versioned migrations workflow works very well wit sqlc. In fact, in most cases, you can simply
```
```suggestion
Atlas's versioned migrations workflow works very well with sqlc. In fact, in most cases, you can simply
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
```diff
@@ -0,0 +1,285 @@
+called "[versioned migration authoring](/versioned/diff)", we still define our desired state in HCL or
+SQL but use Atlas to compute the required changes, storing them in the migration directory as SQL files.
+
+This strategy brings the best of both worlds, we can have a single source of truth (our desired state) and have all the
```
```suggestion
This strategy brings the best of both worlds, where we can have a single source of truth (our desired state) and have all the
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
```diff
@@ -0,0 +1,285 @@
+authored SQL files that will be used to reach the desired state. To learn more about this approach and see it in action,
+watch this [video](https://www.youtube.com/watch?v=L-UlkXtp3OY).
+
+Now that we have a bit of understanding of all the concepts, let's see how we can use Atlas with sqlc, using
```
```suggestion
Now that we have a bit of an understanding of all the concepts, let's see how we can use Atlas with sqlc, using
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
```diff
@@ -0,0 +1,285 @@
+
+## Creating the initial migration directory
+
+The first thing we should do is create the initial migration directory, if you are following
```
```suggestion
The first thing we should do is create the initial migration directory. If you are following
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
````diff
@@ -0,0 +1,285 @@
+  --dev-url "docker://postgres?search_path=public"
+```
+
+A new `migrations` directory should be created with two files on it.
````
```suggestion
A new `migrations` directory should be created with two files in it.
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
```diff
@@ -0,0 +1,285 @@
+Like in the last guide, we need a temporary database to check, simulate and validate the
+generated queries. This database is provided using the `dev-url` flag. Atlas has support for running database
+containers, which is what we are using in this example. Fore more information about dev
+database, [read about it here](/concepts/dev-database).
```
```suggestion
databases, [read about them here](/concepts/dev-database).
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
```diff
@@ -0,0 +1,285 @@
+database, [read about it here](/concepts/dev-database).
+:::
+
+From the sqlc side, we don't have to do anything since have not changed our `schema.sql` or `query.sql` file.
```
```suggestion
From the sqlc side, we don't have to do anything since we have not changed our `schema.sql` or `query.sql` file.
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
```diff
@@ -0,0 +1,285 @@
+## Migrating the database
+
+In this guide, we are using a Postgres
+database running in a local Docker container, you can start such a container by running:
```
```suggestion
database running in a local Docker container. You can start such a container by running:
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
````diff
@@ -0,0 +1,285 @@
+docker run --rm -d --name atlas-sqlc -p 5432:5432 -e POSTGRES_PASSWORD=pass -e POSTGRES_DB=sqlc postgres
+```
+
+Now that we have our base migration directory and target database, we can migrate our schema based on it, to do so we
````
```suggestion
Now that we have our base migration directory and target database, we can migrate our schema based on it. To do so, we
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
````diff
@@ -0,0 +1,285 @@
+```
+
+The command `migrate apply` connects to the given database [URL](/concepts/url) provided in the `url` flag, checks the
+current state of the database and applies any pending migration.
````
```suggestion
current state of the database and applies any pending migrations.
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
````diff
@@ -0,0 +1,285 @@
+```
+
+When working with sqlc, after every change to the `schema.sql` or `query.sql` file, we should execute the sqlc generate
+command again, even in the cases where the changes seens to not cause problems, like adding a new column, this is a
````
```suggestion
command again. Even in the cases where the changes seem to not cause problems, like adding a new column, this is a
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
````diff
@@ -0,0 +1,285 @@
+```
+
+:::note
+sqlc only shows messages when encountering an error, if the generator process occurred successful, no output should be
````
```suggestion
sqlc only shows messages when encountering an error, if the generator process was successful, no output should be
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
````diff
@@ -0,0 +1,285 @@
+  --dev-url "docker://postgres?search_path=public"
+```
+
+If you look at the migration directory, a new SQL file should be created with the commands required for adding a new
````
```suggestion
If you look at the migration directory, a new SQL file should be created with the statements required for adding a new
```

hilakashai commented on doc/md/guides/frameworks/sqlc-versioned.md:
````diff
@@ -0,0 +1,285 @@
+```
+
+If you look at the migration directory, a new SQL file should be created with the commands required for adding a new
+column called `age`.
````
```suggestion
`age` column.
```

hilakashai reviewed (commented):
Great work!!

rotemtam reviewed (approved):
Well done 🚀
